crates/workspace/src/workspace.rs

```diff
@@ -3122,6 +3123,51 @@ impl Workspace {
         }))
     }
 
+    pub fn join_all_panes(&mut self, cx: &mut ViewContext<Self>) {
+        let active_item = self.active_pane.read(cx).active_item().take();
+        for i in 0..self.panes.len() {
+            self.join_pane_into_active(&self.panes[i].clone(), cx);
+        }
+        if let Some(active_item) = active_item {
+            self.activate_item(&(*active_item), true, true, cx);
+        }
+        cx.notify();
+    }
+
+    fn join_pane_into_active(&mut self, pane: &View<Pane>, cx: &mut ViewContext<'_, Workspace>) {
+        if *pane == self.active_pane {
+            return;
+        } else if pane.read(cx).items_len() == 0 {
+            self.close_pane(pane, cx);
+        } else {
+            self.move_all_items(pane, &self.active_pane.clone(), cx);
+        }
+    }
+
+    fn close_pane(&mut self, pane: &View<Pane>, cx: &mut ViewContext<Self>) {
+        pane.update(cx, |_, cx| {
+            cx.emit(pane::Event::Remove {
+                focus_on_pane: None,
+            })
+        })
+    }
+
+    fn move_all_items(
```

Comment:
I'm running into the same issue here as:
https://github.com/zed-industries/zed/pull/12554#discussion_r1623208955

Like the other author, I also welcome advice on how to improve this